### DISCUSSION TITLE
i have problème in the code i dont know wher

### DISCUSSION THREAD

**Discussion by fatiam46** [datasciencecompleteguide/Machine-Learning#23] [General]
**can you help me please**

import cv2
import numpy as np

widthImg = 480
heightImg = 640

cap = cv2.VideoCapture(0)
cap.set(3, widthImg)
cap.set(4, heightImg)
cap.set(10, 150)

kernel = np.ones((5, 5))

def find_sheet_end_point(window):

    window = window.reshape((4, 2))
    new_window = np.zeros((4, 2), dtype=np.float32)
    add = window.sum(axis=1)
    diff = np.diff(window, axis=1)
    new_window[0] = window[np.argmin(add)]
    new_window[1] = window[np.argmin(add)]
    new_window[2] = window[np.argmax(diff)]
    new_window[3] = window[np.argmax(diff)]
    return new_window

def preProcessing(img):
    imgGray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    imgBlur = cv2.GaussianBlur(imgGray, (9, 9), 0)
    imgCanny = cv2.Canny(imgBlur, 28, 48)
    imgDialte = cv2.dilate(imgCanny, kernel, iterations=2)  #deux fois dialation
    imgerrod = cv2.erode(imgDialte, kernel, iterations=1)

    contours, hierarchy = cv2.findContours(imgCanny, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
    contours = sorted(contours, key=cv2.contourArea, reverse=True)

    selected_window = ""
    for contour in contours:
        contour_perimeter = cv2.arcLength(contour, True)
        contour_approximation = cv2.approxPolyDP(contour, 0.02 * contour_perimeter, True)

        if len(contour_approximation) == 4:
            selected_window = contour_approximation
            break
    contour_approximation = find_sheet_end_point(selected_window)
    points = np.float32([[0, 0], [widthImg, 0], [widthImg, heightImg], [0, heightImg]])
    perspective = cv2.getPerspectiveTransform(contour_approximation, points)
    wrap = cv2.warpPerspective(img, perspective, (widthImg, heightImg))

    return wrap

while True:
    veri, img = cap.read()
    img = cv2.resize(img, (widthImg, heightImg))
    imgContour = img.copy()
    imgThres = preProcessing(img)
    cv2.imshow("Resultat", imgThres)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

![Capture d’écran 2021-08-20 181630](https://user-images.githubusercontent.com/87945994/130325153-bd279975-db6d-4b81-b2b3-e817a0c1f8bf.png)

**Comment by datasciencecompleteguide** [comment]
Today I tried to resolve it for some time but till now, I was only able to find one problm and that is insert the line given below just after the find_sheet_end_point function.

def find_sheet_end_point(window):
            window = np.array(window)

I am still trying to get it, will update you asap.
You can also give me any hint if you can.😅

**Reply by fatiam46** [reply]
Thank you, sir, and sorry for the inconvenience.
so in this code im trying to make a document scanner using open cv 
and im still a beginner in python

**Reply by fatiam46** [reply]
on the principal i used this code but he give me an erreur when i added : reorder(biggest) in function : def get wrap
so i tride de modifie the code 
![Capture d’écran 2021-08-21 204827](https://user-images.githubusercontent.com/87945994/130333328-5a686abb-2653-442b-bda7-c648779a64b1.png)



import cv2
import numpy as np


####################
widthImg = 360
heightImg = 640
#######################


cap = cv2.VideoCapture(0)
cap.set(3, 640)
cap.set(4, 480)
cap.set(10, 150)


def preProcessing(img):
    imgGray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    imgBlur = cv2.GaussianBlur(imgGray, (5, 5), 1)
    imgCanny = cv2.Canny(imgBlur, 200, 200)
    kernel = np.ones((5, 5))
    imgDialte = cv2.dilate(imgCanny, kernel, iterations=2)  # deux fois dialation
    imgThres = cv2.erode(imgDialte, kernel, iterations=1)
    return imgThres

def getContours(img):
    biggest = np.array([])
    maxArea = 0
    contours, hierarchy = cv2.findContours(img, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
    for cnt in contours:
        area = cv2.contourArea(cnt)
        if area > 5000:
            #cv2.drawContours(imgContour, cnt, -1, (255, 0, 0), 3)
            peri = cv2.arcLength(cnt, True)
            approx = cv2.approxPolyDP(cnt, 0.02*peri, True)
            if area > maxArea and  len(approx) == 4:
                biggest = approx
                maxArea = area
    cv2.drawContours(imgContour, biggest, -1, (255, 0, 0), 20)
    return biggest


def reorder(myPoints):
    myPoints = np.array(myPoints)
    myPoints =myPoints.reshape((4, 2))
    myPointsNew = np.zeros((4, 1, 2), np.int32)
    add = myPoints.sum(1)
    #print("add", add)

    myPointsNew[0] = myPoints[np.argmin(add)]
    myPointsNew[3] = myPoints[np.argmax(add)]
    diff = np.diff(myPoints, axis=1)
    myPointsNew[1] = myPoints[np.argmin(diff)]
    myPointsNew[2] = myPoints[np.argmax(diff)]
    #print("NewPoints", myPointsNew)
    return myPointsNew


def getWarp(img, biggest):

    reorder(biggest)
    print(biggest)
    pts1 = np.float32(biggest)
    pts1 = np.float32([[0, 0], [widthImg, 0], [0, heightImg], [widthImg, heightImg]])
    matrix = cv2.getPerspectiveTransform(pts1, pts1)
    imgOutput = cv2.warpPerspective(img, matrix, (widthImg, heightImg))

    return imgOutput


while True:
    success, img = cap.read()
    img = cv2.resize(img, (widthImg, heightImg))
    imgContour = img.copy()

    imgThres = preProcessing(img)
    biggest = getContours(imgThres)
    #print(biggest)

    imgWarped = getWarp(img, biggest)

    cv2.imshow("Result", imgWarped)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

**Comment by datasciencecompleteguide** [comment]
def getContours(img):
    biggest = np.array([])
    maxArea = 0
    contours, hierarchy = cv2.findContours(img, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
    for cnt in contours:
        area = cv2.contourArea(cnt)
        if area > 5000:
            #cv2.drawContours(imgContour, cnt, -1, (255, 0, 0), 3)
            peri = cv2.arcLength(cnt, True)
            approx = cv2.approxPolyDP(cnt, 0.02*peri, True)
            if area > maxArea and len(approx) == 4:
                biggest = approx
                maxArea = area
                cv2.drawContours(imgContour, biggest, -1, (255, 0, 0), 20)
    return biggest

1. In this function the threshold values that you are taking for the area is too large (instead try with 100 something nearby it...you can check it by simply printing the values of area).

2. Next, thing is that after the inner if statement i.e., (if area>maxarea and len(approx)==4), you are trying to draw the controus and when you are returning the biggest than every time its shape is something between (4, 1, 2) to (12, 1, 2). So, you need to change the conditions there also.

3. Afrer returning the biggest or before returning it use the line: (First, resolve the above two errors)
       biggest =  np.array(biggest, dtype=np.uint8) (to convert it to the matrix)

You can also contact me on Instagram:
[CrazzyLearners](https://www.instagram.com/crazzylearners/?hl=en)
